codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -101,89 +79,132 @@ public Http2MultiplexCodec(boolean server, ChannelHandler streamHandler) {
-            new ConnectionListener().onStreamActive(stream);
-            upgrade.upgradeRequest().headers().setInt(
-                HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
-            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
-                .channelRead(ctx, upgrade.upgradeRequest().retain());
+            Http2StreamChannel childChannel = childChannels.get(streamEx.streamId());
+            if (childChannel != null) {
+                childChannel.pipeline().fireExceptionCaught(streamEx);
+            }
         } finally {
-            upgrade.release();
+            onStreamClosed(streamEx.streamId());
         }
     }
 
     // Override this to signal it will never throw an exception.
     @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        ctx.fireExceptionCaught(cause);
+    public void flush(ChannelHandlerContext ctx) {
+        ctx.flush();
     }
 
-    // Override this to signal it will never throw an exception.
     @Override
-    public void flush(ChannelHandlerContext ctx) {
-        ctx.flush();
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof Http2Frame)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+        if (msg instanceof Http2StreamFrame) {
+            Http2StreamFrame frame = (Http2StreamFrame) msg;
+            int streamId = frame.streamId();
+            Http2StreamChannel childChannel = childChannels.get(streamId);
+            if (childChannel == null) {
+                ReferenceCountUtil.release(msg);
+                throw new StreamException(streamId, STREAM_CLOSED, format("Received %s frame for an unknown stream %d",
+                                                                          frame.name(), streamId));
+            }
+            fireChildReadAndRegister(childChannel, frame);
+        } else if (msg instanceof Http2GoAwayFrame) {
+            Http2GoAwayFrame goAwayFrame = (Http2GoAwayFrame) msg;
+            for (PrimitiveEntry<Http2StreamChannel> entry : childChannels.entries()) {
+                Http2StreamChannel childChannel = entry.value();
+                int streamId = entry.key();
+                if (streamId > goAwayFrame.lastStreamId() && isLocalStream(streamId)) {
+                    childChannel.pipeline().fireUserEventTriggered(goAwayFrame.retainedDuplicate());
+                }
+            }
+            goAwayFrame.release();
+        } else {
+            throw new UnsupportedMessageTypeException(msg);
+        }
+    }
+
+    void fireChildReadAndRegister(Http2StreamChannel childChannel, Http2StreamFrame frame) {
+        // Can't use childChannel.fireChannelRead() as it would fire independent of whether
+        // channel.read() had been called.
+        childChannel.fireChildRead(frame);
```

Comment:
actually it looks like `AbstractHttp2StreamChannel.readInProgress` should account for the `read()` being called ... and will queue otherwise ... maybe remove/clarify the above comment?